sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -327,24 +359,28 @@ class CodegenContext {
       nullable: Boolean,
       isVectorized: Boolean = false): String = {
     if (nullable) {
-      // Can't call setNullAt on DecimalType, because we need to keep the offset
-      if (!isVectorized && dataType.isInstanceOf[DecimalType]) {
+      // Can't call setNullAt for mutable but non-primitive data type, e.g. DecimalType, StructType
+      // with fix-length fields, because we need to keep the offset and length.
+      val isMutableVarLenField = UnsafeRow.isMutable(dataType) && !UnsafeRow.isFixedLength(dataType)
```

Comment:
Do you think it is good to add `isMutableVarLenField` method into `UnsafeRow` directly?